File: android/app/src/main/java/com/molt/marketplace/core/designsystem/component/MoltCard.kt
package com.molt.marketplace.core.designsystem.component

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import com.molt.marketplace.R
import com.molt.marketplace.core.designsystem.theme.MoltCornerRadius
import com.molt.marketplace.core.designsystem.theme.MoltElevation
import com.molt.marketplace.core.designsystem.theme.MoltSpacing
import com.molt.marketplace.core.designsystem.theme.MoltTheme

@Composable
@Suppress("ktlint:standard:function-naming", "CognitiveComplexMethod")
fun MoltProductCard(
    imageUrl: String?,
    title: String,
    price: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    originalPrice: String? = null,
    vendorName: String? = null,
    rating: Float? = null,
    reviewCount: Int? = null,
    isWishlisted: Boolean = false,
    onWishlistToggle: (() -> Unit)? = null,
) {
    Card(
        modifier = modifier
            .fillMaxWidth()
            .clickable(onClick = onClick),
        shape = RoundedCornerShape(MoltCornerRadius.Medium),
        elevation = CardDefaults.cardElevation(defaultElevation = MoltElevation.Level1),
    ) {
        Column {
            Box {
                MoltImage(
                    url = imageUrl,
                    contentDescription = title,
                    modifier = Modifier
                        .fillMaxWidth()
                        .aspectRatio(16f / 9f),
                )

                if (onWishlistToggle != null) {
                    val wishlistDescription = if (isWishlisted) {
                        stringResource(R.string.common_remove_from_wishlist)
                    } else {
                        stringResource(R.string.common_add_to_wishlist)
                    }

                    IconButton(
                        onClick = onWishlistToggle,
                        modifier = Modifier.align(Alignment.TopEnd),
                    ) {
                        Icon(
                            imageVector = if (isWishlisted) {
                                Icons.Filled.Favorite
                            } else {
                                Icons.Outlined.FavoriteBorder
                            },
                            contentDescription = wishlistDescription,
                            tint = if (isWishlisted) {
                                MaterialTheme.colorScheme.error
                            } else {
                                MaterialTheme.colorScheme.onSurface
                            },
                        )
                    }
                }
            }

            Column(modifier = Modifier.padding(MoltSpacing.CardPadding)) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleMedium,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                )

                if (vendorName != null) {
                    Spacer(modifier = Modifier.height(MoltSpacing.XS))
                    Text(
                        text = vendorName,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }

                Spacer(modifier = Modifier.height(MoltSpacing.XS))
                MoltPriceText(
                    price = price,
                    originalPrice = originalPrice,
                )

                if (rating != null) {
                    Spacer(modifier = Modifier.height(MoltSpacing.XS))
                    MoltRatingBar(
                        rating = rating,
                        showValue = true,
                        reviewCount = reviewCount,
                    )
                }
            }
        }
    }
}

@Composable
@Suppress("ktlint:standard:function-naming", "CognitiveComplexMethod")
fun MoltInfoCard(
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    leadingIcon: ImageVector? = null,
    trailingContent: @Composable (() -> Unit)? = null,
    onClick: (() -> Unit)? = null,
) {
    Card(
        modifier = modifier
            .fillMaxWidth()
            .then(
                if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier,
            ),
        shape = RoundedCornerShape(MoltCornerRadius.Medium),
        elevation = CardDefaults.cardElevation(defaultElevation = MoltElevation.Level1),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MoltSpacing.CardPadding),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(MoltSpacing.MD),
        ) {
            if (leadingIcon != null) {
                Icon(
                    imageVector = leadingIcon,
                    contentDescription = null,
                    modifier = Modifier.size(MoltSpacing.LG),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleMedium,
                )
                if (subtitle != null) {
                    Spacer(modifier = Modifier.height(MoltSpacing.XS))
                    Text(
                        text = subtitle,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }

            if (trailingContent != null) {
                trailingContent()
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
@Suppress("ktlint:standard:function-naming")
private fun MoltProductCardPreview() {
    MoltTheme {
        MoltProductCard(
            imageUrl = null,
            title = "Premium Wireless Headphones with Noise Cancellation",
            price = "29.99",
            originalPrice = "39.99",
            vendorName = "TechStore",
            rating = 4.5f,
            reviewCount = 123,
            isWishlisted = false,
            onWishlistToggle = {},
            onClick = {},
        )
    }
}

@Preview(showBackground = true)
@Composable
@Suppress("ktlint:standard:function-naming")
private fun MoltProductCardWishlistedPreview() {
    MoltTheme {
        MoltProductCard(
            imageUrl = null,
            title = "Simple Product",
            price = "9.99",
            isWishlisted = true,
            onWishlistToggle = {},
            onClick = {},
        )
    }
}

@Preview(showBackground = true)
@Composable
@Suppress("ktlint:standard:function-naming")
private fun MoltInfoCardPreview() {
    MoltTheme {
        MoltInfoCard(
            title = "Shipping Address",
            subtitle = "123 Main St, Valletta",
        )
    }
}

File: ios/XiriGoEcommerce/Core/DesignSystem/Component/MoltCard.swift
import SwiftUI

// MARK: - MoltProductCard

struct MoltProductCard: View {
    // MARK: - Properties

    private let imageUrl: URL?
    private let title: String
    private let price: String
    private let originalPrice: String?
    private let vendorName: String?
    private let rating: Double?
    private let reviewCount: Int?
    private let isWishlisted: Bool
    private let onWishlistToggle: (() -> Void)?
    private let action: () -> Void

    // MARK: - Init

    init(
        imageUrl: URL?,
        title: String,
        price: String,
        originalPrice: String? = nil,
        vendorName: String? = nil,
        rating: Double? = nil,
        reviewCount: Int? = nil,
        isWishlisted: Bool = false,
        onWishlistToggle: (() -> Void)? = nil,
        action: @escaping () -> Void
    ) {
        self.imageUrl = imageUrl
        self.title = title
        self.price = price
        self.originalPrice = originalPrice
        self.vendorName = vendorName
        self.rating = rating
        self.reviewCount = reviewCount
        self.isWishlisted = isWishlisted
        self.onWishlistToggle = onWishlistToggle
        self.action = action
    }

    // MARK: - Body

    var body: some View {
        Button(action: action) {
            VStack(alignment: .leading, spacing: MoltSpacing.xs) {
                imageSection
                contentSection
            }
            .background(MoltColors.surface)
            .clipShape(RoundedRectangle(cornerRadius: MoltCornerRadius.medium))
            .moltElevation(MoltElevation.level1)
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(accessibilityDescription)
        .accessibilityAddTraits(.isButton)
    }

    // MARK: - Subviews

    private var imageSection: some View {
        ZStack(alignment: .topTrailing) {
            MoltImage(url: imageUrl)
                .aspectRatio(16.0 / 9.0, contentMode: .fill)
                .clipped()

            if let onWishlistToggle {
                Button(action: onWishlistToggle) {
                    Image(systemName: isWishlisted ? "heart.fill" : "heart")
                        .foregroundStyle(isWishlisted ? MoltColors.error : MoltColors.onSurfaceVariant)
                        .font(.system(size: MoltSpacing.IconSize.medium))
                        .padding(MoltSpacing.sm)
                        .background(MoltColors.surface.opacity(0.8))
                        .clipShape(Circle())
                }
                .frame(minWidth: MoltSpacing.minTouchTarget, minHeight: MoltSpacing.minTouchTarget)
                .padding(MoltSpacing.xs)
                .accessibilityLabel(
                    isWishlisted
                        ? String(localized: "common_remove_from_wishlist")
                        : String(localized: "common_add_to_wishlist")
                )
            }
        }
    }

    private var contentSection: some View {
        VStack(alignment: .leading, spacing: MoltSpacing.xs) {
            Text(title)
                .font(MoltTypography.titleMedium)
                .foregroundStyle(MoltColors.onSurface)
                .lineLimit(2)

            if let vendorName {
                Text(vendorName)
                    .font(MoltTypography.bodySmall)
                    .foregroundStyle(MoltColors.onSurfaceVariant)
            }

            MoltPriceText(price: price, originalPrice: originalPrice)

            if let rating {
                MoltRatingBar(
                    rating: rating,
                    showValue: true,
                    reviewCount: reviewCount
                )
            }
        }
        .padding(MoltSpacing.cardPadding)
    }

    // MARK: - Accessibility

    private var accessibilityDescription: String {
        var parts = [title, price]
        if let vendorName {
            parts.append(vendorName)
        }
        if let rating {
            parts.append(
                String(localized: "common_rating_description \(rating) \(5)")
            )
        }
        return parts.joined(separator: ", ")
    }
}

// MARK: - MoltInfoCard

struct MoltInfoCard<TrailingContent: View>: View {
    // MARK: - Properties

    private let title: String
    private let subtitle: String?
    private let leadingIcon: String?
    private let trailingContent: TrailingContent?
    private let action: (() -> Void)?

    // MARK: - Init

    init(
        title: String,
        subtitle: String? = nil,
        leadingIcon: String? = nil,
        action: (() -> Void)? = nil,
        @ViewBuilder trailingContent: () -> TrailingContent
    ) {
        self.title = title
        self.subtitle = subtitle
        self.leadingIcon = leadingIcon
        self.action = action
        self.trailingContent = trailingContent()
    }

    // MARK: - Body

    var body: some View {
        let content = HStack(spacing: MoltSpacing.md) {
            if let leadingIcon {
                Image(systemName: leadingIcon)
                    .foregroundStyle(MoltColors.onSurfaceVariant)
                    .font(.system(size: MoltSpacing.IconSize.medium))
                    .accessibilityHidden(true)
            }

            VStack(alignment: .leading, spacing: MoltSpacing.xxs) {
                Text(title)
                    .font(MoltTypography.titleMedium)
                    .foregroundStyle(MoltColors.onSurface)

                if let subtitle {
                    Text(subtitle)
                        .font(MoltTypography.bodySmall)
                        .foregroundStyle(MoltColors.onSurfaceVariant)
                }
            }

            Spacer()

            if let trailingContent {
                trailingContent
            }
        }
        .padding(MoltSpacing.cardPadding)
        .background(MoltColors.surface)
        .clipShape(RoundedRectangle(cornerRadius: MoltCornerRadius.medium))
        .moltElevation(MoltElevation.level1)

        if let action {
            Button(action: action) {
                content
            }
            .buttonStyle(.plain)
            .accessibilityAddTraits(.isButton)
        } else {
            content
        }
    }
}

extension MoltInfoCard where TrailingContent == EmptyView {
    init(
        title: String,
        subtitle: String? = nil,
        leadingIcon: String? = nil,
        action: (() -> Void)? = nil
    ) {
        self.title = title
        self.subtitle = subtitle
        self.leadingIcon = leadingIcon
        self.action = action
        self.trailingContent = nil
    }
}

// MARK: - Previews

#Preview("MoltProductCard") {
    MoltProductCard(
        imageUrl: nil,
        title: "Premium Wireless Headphones with Noise Cancellation",
        price: "29.99",
        originalPrice: "39.99",
        vendorName: "TechStore",
        rating: 4.5,
        reviewCount: 123,
        isWishlisted: false,
        onWishlistToggle: {},
        action: {}
    )
    .frame(width: 200)
    .padding()
}

#Preview("MoltProductCard Wishlisted") {
    MoltProductCard(
        imageUrl: nil,
        title: "Simple Product",
        price: "9.99",
        isWishlisted: true,
        onWishlistToggle: {},
        action: {}
    )
    .frame(width: 200)
    .padding()
}

#Preview("MoltInfoCard") {
    VStack(spacing: MoltSpacing.sm) {
        MoltInfoCard(
            title: "Shipping Address",
            subtitle: "123 Main Street, Valletta",
            leadingIcon: "location",
            action: {}
        )

        MoltInfoCard(
            title: "Payment Method",
            subtitle: "Visa ending in 4242",
            leadingIcon: "creditcard"
        ) {
            Image(systemName: "chevron.right")
                .foregroundStyle(MoltColors.onSurfaceVariant)
        }
    }
    .padding()
}
